README.md
MapCaller: An efficient and versatile approach for short-read mapping and variant identification using high-throughput sequenced data.
===================

Developers: Dr. Hsin-Nan Lin and Dr. Wen-Lian Hsu. Institute of Information Science, Academia Sinica, Taiwan.

# Introduction
MapCaller aligns every NGS short read against a reference genome and collects all the alignment information to deduce sequence variants. MapCaller adopts the mapping algorithm of KART to perform read alignment. It maintains a position frequency matrix to keep track of every nucleotide¡¦s frequency at each position in the reference genome, and it collects all insertion and deletion events which are found during the read mapping. Furthermore, MapCaller also learns all possible break points from discordant or partial read alignments. Finally, MapCaller identifies sequence variants based on all the above-mentioned information. The novelty of our algorithm derives from the integration of read mapping and the variation information gathering into a coherent system for genomic variant calling. Thus, the inputs to MapCaller is one or more NGS read file and an index file for the reference genome, and the output is a VCF file for the variant calling result.

For more information, please refer to our manuscript (https://www.biorxiv.org/content/10.1101/783605v2)

The benchmark data sets and the performance evaluation program could be found at http://bioapp.iis.sinica.edu.tw/~arith/MapCaller/

# Installation

## Conda
Install [Bioconda](https://bioconda.github.io/user/install.html) then type:
```
$ conda install -c conda-forge -c bioconda -c defaults mapcaller
```

## Homebrew
Install [HomeBrew](http://brew.sh/) (MacOS) or [LinuxBrew](http://linuxbrew.sh/) (Linux) then type:
```
$ brew install brewsci/bio/mapcaller
```

# Download

Please use the command 
  ```
  $ git clone https://github.com/hsinnan75/MapCaller.git
  ```
to download the package of MapCaller.

# Dependencies
To compile MapCaller, it requires `libboost-all-dev`, `libbz2-dev`, and `liblzma-dev` pre-installed in your system.

# Compiling
To compile MapCaller and the index tool, please just type `make` to compile MapCaller. If the compilation or the program fails, please contact me (arith@iis.sinica.edu.tw) or report it at the
[Issue Tracker](https://github.com/hsinnan75/MapCaller/issues).

# Test
You may run `run_test.sh` to test MapCaller with a toy example.

# Get updates
  ```
  $ bin/MapCaller update
  ```
or
  ```
  $ git fetch
  $ git merge origin/master master
  $ make
  ```

# Instructions

  ```
  $ bin/MapCaller [options]
  ```

# Usage

To index a reference genome, it requires the target genome file (in fasta format) and the prefix of the index files (including the directory path).

  ```
  $ bin/MapCaller index ref_file[ex.ecoli.fa] index_prefix[ex. Ecoli]
  ```
The above command is to index the genome file and store the index files begining with $index_prefix.

or

  ```
  $ bin/MapCaller -r ref_file[ex.ecoli.fa] 
  ```
The above command is to read a reference file and build a temporary index file

To perform variant calling, MapCaller requires the the index files of the reference genome and at least one read file (two read files for the separated paired-end reads). Users should use -i to specify the prefix of the index files (including the directory path).

 case 1: standard vcf output / sam output (optional) / bam output (optional)
  ```
 $ bin/MapCaller -i ecoli -f ReadFile1.fa -f2 ReadFile2.fa -vcf out.vcf [-sam out.sam][-bam out.bam]
  ```

 case 2: multiple input 
  ```
 $ bin/MapCaller -i ecoli -f ReadFileA_1.fq ReadFileB_1.fq ReadFileC_1.fq -f2 ReadFileA_2.fq ReadFileB_2.fq ReadFileC_2.fq -vcf out.vcf
  ```

 case 3: given a reference genome
  ```
 $ bin/MapCaller -r ecoli.fa -f ReadFile1.fa -f2 ReadFile2.fa -vcf out.vcf [-sam out.sam][-bam out.bam]
  ```

# File formats

- Reference genome files

    All reference genome files should be in FASTA format.

- Read files

    MapCaller is designed for NGS short reads. All reads files should be in FASTA/FASTQ format. Input files can be compressed with gzip format.
    Please note, if reads are stored in FASTA format, each read sequence is not allowed to be wrapped (split over multiple lines).
    Read sequences should be capital letters. The quality scores in FASTQ are not considered in the alignments. 
    If paired-end reads are separated into two files, use -f and -f2 to indicate the two filenames. The i-th reads in the two files are paired. If paired-end reads are in the same file, use -p. The first and second reads are paired, the third and fourth reads are paired, and so on. For the latter case, use -p to indicate the input file contains paired-end reads.

- Output file

    MapCaller outputs a SAM/BAM file [optional] and a VCF file. 

# Parameter setting

 ```
-t INT number of threads [16]

-i STR index prefix [optional, BWT based (BWA)]

-r STR reference filename [optional, fasta]

-f STR read filename [required, fasta or fastq or fq.gz]

-f2 STR read filename2 [optional, fasta or fastq or fq.gz], f and f2 are files with paired reads

-p the input read file consists of interleaved paired-end sequences [false]

-sam STR SAM output [optional, default: no mapping output]

-bam STR BAM output [optional, default: no mapping output]

-alg STR gapped alignment algorithm [optional, nw|ksw2, default: nw]

-vcf STR VCF output [output.vcf]

-dup INT Maximal PCR duplicates [optional, 1-100, default: 5]

-ploidy INT number of sets of chromosomes in a cell [optional, default:2]

-filter Apply variant filters (under test) [false]

-no_vcf No VCF output [false]

-gvcf GVCF mode [false]

-size Sequencing fragment size [default: 500, MapCaller can predict the fragment size automatically]

-ad INT Minimal ALT allele count [3]

-somatic detect somatic mutations [false]

-m output multiple alignments [false]

-v version number

-h help

  ```
# Acknowledgements
We would like to thank [A/Prof. Torsten Seemann](https://github.com/tseemann)
and [Dr. Devon Ryan](https://github.com/dpryan79) for their help with the software.

# Note
The source code of the homemade genome sequence variation simulator is available at src/sv_simulator/.


CHANGES.md
# Changes
- 0.9.9.1: Adjusted read depth threshold.
- 0.9.9.2: Removed variants which appear in repetitive regions.
- 0.9.9.3: Adjusted read depth threshold for somatic mutation detection.
- 0.9.9.4: Added variant filters.
- 0.9.9.5: Fixed a bug on read count.
- 0.9.9.6: Fixed a bug on read count.
- 0.9.9.7: Fixed typos and warnings.
- 0.9.9.8: Used ksw2 algorithm for gapped alignments and fixed a bug on VCF format output.
- 0.9.9.9: Added a new argument(-ploidy) for setting different ploidy.
- 0.9.9.a: Fixed a bug on read mapping && Added an option (-dup) to filter out PCR-duplicates.
- 0.9.9.b: Added GVCF format (-gvcf).
- 0.9.9.c: Added a new argument(-monomorphic) to report all loci which do not have any potential alternates.
- 0.9.9.d: Fixed bugs on checking reference nucloetides and reporting deletion events.
- 0.9.9.e: Fixed bugs on reporting monomorphic regions and reported unmapped regions (potential deletions).
- 0.9.9.15: Fixed errors on VCF headers.
- 0.9.9.16: Added a new argument (-maxmm) for setting maximal mismatches in read alignment.
- 0.9.9.17: Updated VCF flags and format.
- 0.9.9.18: Fixed bugs && Added a new argument (-maxclip) for setting maximal clip size in read alignment.
- 0.9.9.19: Fixed a bug on the determination of deletion locus.
- 0.9.9.21: Reported CNV (N>=2) regions with the flag "DUP" (experimental).
- 0.9.9.22: Change the argument "-maxmm" to the maximal mismatch rate allowed in a read alignment.
- 0.9.9.23: Change the VCF output to fix the problem when applying vcftools with --plink
- 0.9.9.24: Summarize the process to a log file
- 0.9.9.25: Fixed typos in vcf headers.
- 0.9.9.26: Fixed a bug when running multi-threads on MacOS.
- 0.9.9.27: Integrated bwt_index into MapCaller
- 0.9.9.28: Build reference index on the fly
- 0.9.9.29: Fixed a bug when the input data contains unexpected characters.
- 0.9.9.30: Fixed a bug on buiding index
- 0.9.9.31: Modified the definition of tag AD in VCF output
- 0.9.9.32: Added the information of F1R2/F2R1 counts in VCF files.
- 0.9.9.33: Fixed a bug on reporting del events.
- 0.9.9.34: Improved the quality measure of variant called.


src/main.cpp
#include "structure.h"

#ifdef __cplusplus
extern "C"
{
	int bwa_idx_build(const char *fa, const char *prefix);
}
#endif

bwt_t *Refbwt;
bwaidx_t *RefIdx;
const char* VersionStr = "0.9.9.34";

string CmdLine;
int iChromsomeNum;
uint8_t iMaxDuplicate;
time_t StartProcessTime;
map<string, int> ChrIdMap;
map<int64_t, int> PosChrIdMap;
map<int64_t, bool> KnowSiteMap;
int64_t GenomeSize, TwoGenomeSize;
vector<Chromosome_t> ChromosomeVec;
float FrequencyThr, MaxMisMatchRate;
MappingRecord_t* MappingRecordArr = NULL;
vector<string> ReadFileNameVec1, ReadFileNameVec2;
int64_t ObservGenomicPos, ObserveBegPos, ObserveEndPos;
pthread_mutex_t LibraryLock, ProfileLock, OutputLock, VarLock;
char *RefSequence, *RefFileName, *KnownSiteFileName, *IndexFileName, *SamFileName, *VcfFileName, *LogFileName;
int iThreadNum, MaxPosDiff, iPloidy, FragmentSize, MaxClipSize, MinAlleleDepth, MinIndFreq, MinVarConfScore, MinCNVsize, MinUnmappedSize;
bool bDebugMode, bFilter, bPairEnd, bUnique, bSAMoutput, bSAMFormat, bGVCF, bMonomorphic, bVCFoutput, bSomatic, gzCompressed, FastQFormat, NW_ALG;

void ShowProgramUsage(const char* program)
{
	fprintf(stderr, "MapCaller v%s\n\n", VersionStr);
	fprintf(stderr, "Usage: %s -i Index_Prefix -f <ReadFile_A1 ReadFile_B1 ...> [-f2 <ReadFile_A2 ReadFile_B2 ...>]\n\n", program);
	fprintf(stderr, "Options: -i STR        BWT_Index_Prefix\n");
	fprintf(stderr, "         -r STR        Reference filename (format:fa)\n");
	fprintf(stderr, "         -f            files with #1 mates reads (format:fa, fq, fq.gz)\n");
	fprintf(stderr, "         -f2           files with #2 mates reads (format:fa, fq, fq.gz)\n");
	fprintf(stderr, "         -t INT        number of threads [%d]\n", iThreadNum);
	fprintf(stderr, "         -size         sequencing fragment size [%d]\n", FragmentSize);
	fprintf(stderr, "         -indel INT	maximal indel size [%d]\n", MaxPosDiff);
	fprintf(stderr, "         -ad INT       minimal ALT allele count [%d]\n", MinAlleleDepth);
	fprintf(stderr, "         -dup INT      maximal PCR duplicates [%d]\n", iMaxDuplicate);
	fprintf(stderr, "         -maxmm FLOAT  maximal mismatch rate in read alignment [%.2f]\n", MaxMisMatchRate);
	fprintf(stderr, "         -maxclip INT  maximal clip size at either ends [%d]\n", MaxClipSize);
	fprintf(stderr, "         -sam          SAM output filename [NULL]\n");
	fprintf(stderr, "         -bam          BAM output filename [NULL]\n");
	fprintf(stderr, "         -alg STR      gapped alignment algorithm (option: nw|ksw2)\n");
	fprintf(stderr, "         -vcf          VCF output filename [%s]\n", VcfFileName);
	fprintf(stderr, "         -gvcf         GVCF mode [false]\n");
	fprintf(stderr, "         -log STR      log filename [%s]\n", LogFileName);
	fprintf(stderr, "         -monomorphic  report all loci which do not have any potential alternates.\n");
	fprintf(stderr, "         -min_cnv INT  the minimal cnv size to be reported [%d].\n", MinCNVsize);
	fprintf(stderr, "         -min_gap INT  the minimal gap(unmapped) size to be reported [%d].\n", MinUnmappedSize);
	fprintf(stderr, "         -ploidy INT   number of sets of chromosomes in a cell (1:monoploid, 2:diploid) [%d]\n", iPloidy);
	fprintf(stderr, "         -m            output multiple alignments\n");
	fprintf(stderr, "         -somatic      detect somatic mutations [false]\n");
	fprintf(stderr, "         -no_vcf       No VCF output [false]\n");
	fprintf(stderr, "         -p            paired-end reads are interlaced in the same file\n");
	fprintf(stderr, "         -filter       apply variant filters (under test) [false]\n");
	fprintf(stderr, "         -v            version\n");
	fprintf(stderr, "\n");
}

string MakeRefIdx(char* RefFileName)
{
	string str;
	srand((unsigned int)time(NULL)); str.resize(10);
	for (int i = 0; i < 10; i++) str[i] = (unsigned short)(rand() >> 3) % 26 + 97;

	return str;
}

bool CheckOutputFileName(char *FileName)
{
	struct stat s;
	bool bRet = true;
	int i, len = strlen(FileName);

	if (strcmp(FileName, "-") != 0)
	{
		for (i = 0; i < len; i++)
		{
			if (isalnum(FileName[i]) || FileName[i] == '/' || FileName[i] == '.' || FileName[i] == '_' || FileName[i] == '-');
			else
			{
				bRet = false;
				fprintf(stderr, "Warning: [%s] is not a valid filename!\n", FileName);
				break;
			}
		}
		if (stat(FileName, &s) == 0)
		{
			if (s.st_mode & S_IFDIR)
			{
				bRet = false;
				fprintf(stderr, "Warning: %s is a directory!\n", FileName);
			}
		}
	}
	return bRet;
}

bool CheckInputFiles(vector<string>& ReadFileNameVec)
{
	struct stat s;
	string filetype;
	bool bRet = true;

	for (vector<string>::iterator iter = ReadFileNameVec.begin(); iter != ReadFileNameVec.end(); iter++)
	{
		if (stat(iter->c_str(), &s) == -1)
		{
			bRet = false;
			fprintf(stderr, "Cannot access file:[%s]\n", (char*)iter->c_str());
			break;
		}
		else
		{
			filetype = iter->substr(iter->find_last_of('.') + 1);
			for (string::iterator ii = filetype.begin(); ii != filetype.end(); ii++) *ii = tolower(*ii);
			if (filetype != "fq" && filetype != "fa" && filetype != "fastq" && filetype != "fasta" && filetype != "gz")
			{
				bRet = false;
				fprintf(stderr, "Wrong file type:[%s]\n", iter->c_str());
				break;
			}
		}
	}
	return bRet;
}

void ReadLibInput(const char* LibFileName)
{
	fstream file;
	stringstream ss;
	string str, fn1, fn2;

	file.open(LibFileName, ios_base::in);
	while (!file.eof())
	{
		getline(file, str); if (str == "") break;
		if (str[0] == '#') continue;
		ss.clear(); ss.str(str); ss >> fn1 >> fn2;
		if (fn1 != "") ReadFileNameVec1.push_back(fn1);
		if (fn2 != "") ReadFileNameVec2.push_back(fn2);
	}
	file.close();
}

int main(int argc, char* argv[])
{
	int i;
	string parameter, str, random_prefix;

	bGVCF = false;
	iPloidy = 2;
	iThreadNum = 16;
	bPairEnd = false;
	bDebugMode = false;
	bUnique = true;
	bFilter = false;
	NW_ALG = true;
	FastQFormat = true;
	bSAMoutput = false;
	bSAMFormat = true;
	bSomatic = false;
	bVCFoutput = true;
	gzCompressed = false;
	bMonomorphic = false;

	MinIndFreq = 5;
	MaxClipSize = 5;
	MinCNVsize = 50;
	MaxPosDiff = 30;
	iMaxDuplicate = 5;
	FragmentSize = 500;
	MinAlleleDepth = 5;
	FrequencyThr = 0.2;
	MinVarConfScore = 10;
	MinUnmappedSize = 50;
	MaxMisMatchRate = 0.05;
	LogFileName = (char*)"job.log";
	VcfFileName = (char*)"output.vcf";
	ObservGenomicPos = ObserveBegPos = ObserveEndPos = -1;
	RefSequence = RefFileName = IndexFileName = SamFileName = KnownSiteFileName = NULL;

	if (argc == 1 || strcmp(argv[1], "-h") == 0) ShowProgramUsage(argv[0]);
	else if (strcmp(argv[1], "update") == 0)
	{
		system("git fetch;git merge origin/master master;make");
		exit(0);
	}
	else if (strcmp(argv[1], "index") == 0)
	{
		if(argc == 4) bwa_idx_build(argv[2], argv[3]);
		else
		{
			fprintf(stderr, "usage: %s index ref.fa prefix\n", argv[0]);
		}
		exit(0);
	}
	else
	{
		for (CmdLine = argv[0], i = 1; i < argc; i++) CmdLine += " " + (string)argv[i];

		for (i = 1; i < argc; i++)
		{
			parameter = argv[i];

			if (parameter == "-i" && i + 1 < argc) IndexFileName = argv[++i];
			else if (parameter == "-r" && i + 1 < argc) RefFileName = argv[++i];
			else if (parameter == "-f")
			{
				while (++i < argc && argv[i][0] != '-') ReadFileNameVec1.push_back(argv[i]);
				i--;
			}
			else if (parameter == "-f2")
			{
				while (++i < argc && argv[i][0] != '-') ReadFileNameVec2.push_back(argv[i]);
				i--;
			}
			else if (parameter == "-lib" && i + 1 < argc)
			{
				ReadLibInput(argv[++i]);
			}
			else if (parameter == "-t" && i + 1 < argc)
			{
				if ((iThreadNum = atoi(argv[++i])) <= 0)
				{
					fprintf(stderr, "Warning! The thread number should be positive!\n");
					iThreadNum = 4;
				}
			}
			else if (parameter == "-dup" && i + 1 < argc)
			{
				if (atoi(argv[++i]) <= 15) iMaxDuplicate = (int8_t)atoi(argv[i]);
				else fprintf(stderr, "Warning! The PCR-duplicate range is [1-15]!\n");
			}
			else if (parameter == "-filter") bFilter = true;
			else if (parameter == "-size" && i + 1 < argc) FragmentSize = atoi(argv[++i]);
			else if (parameter == "-indel" && i + 1 < argc)
			{
				if ((MaxPosDiff = atoi(argv[++i])) > 100)
				{
					MaxPosDiff = 100;
					fprintf(stderr, "Warning! The maximal indel size is 100!\n");
				}
			}
			else if (parameter == "-ad" && i + 1 < argc) MinAlleleDepth = atoi(argv[++i]);
			else if (parameter == "-ind" && i + 1 < argc) MinIndFreq = atoi(argv[++i]);
			else if (parameter == "-ploidy" && i + 1 < argc)
			{
				if ((iPloidy = atoi(argv[++i])) > 2)
				{
					iPloidy = 2;
					fprintf(stderr, "Warning! MapCaller only supports monoploid and diploid!\n");
				}
			}
			//else if (parameter == "-known_sites" && i + 1 < argc)
			//{
			//	KnownSiteFileName = argv[++i];
			//}
			else if ((parameter == "-sam") && i + 1 < argc)
			{
				bSAMoutput = true;
				bSAMFormat = true;
				SamFileName = argv[++i];
			}
			else if ((parameter == "-bam") && i + 1 < argc)
			{
				bSAMoutput = true;
				bSAMFormat = false;
				SamFileName = argv[++i];
			}
			else if (parameter == "-log" && i + 1 < argc) LogFileName = argv[++i];
			else if ((parameter == "-alg") && i + 1 < argc)
			{
				str = argv[++i];
				if (str == "ksw2") NW_ALG = false;
				else NW_ALG = true; //nw
			}
			else if (parameter == "-maxmm" && i + 1 < argc) MaxMisMatchRate = atof(argv[++i]);
			else if (parameter == "-maxclip" && i + 1 < argc) MaxClipSize = atoi(argv[++i]);
			else if (parameter == "-min_gap" && i + 1 < argc) MinUnmappedSize = atoi(argv[++i]);
			else if (parameter == "-vcf" && i + 1 < argc) VcfFileName = argv[++i];
			else if (parameter == "-gvcf") bGVCF = true;
			else if (parameter == "-monomorphic") bMonomorphic = true;
			else if (parameter == "-no_vcf") bVCFoutput = false;
			else if (parameter == "-somatic") bSomatic = true;
			else if (parameter == "-pair" || parameter == "-p") bPairEnd = true;
			else if (parameter == "-obs" && i + 1 < argc) ObservGenomicPos = atoi(argv[++i]);
			else if (parameter == "-obr" && i + 2 < argc)
			{
				ObserveBegPos = atoi(argv[++i]);
				ObserveEndPos = atoi(argv[++i]);
				fprintf(stderr, "obr[%lld - %lld]\n", (long long) ObserveBegPos, (long long)ObserveEndPos);
			}
			else if (parameter == "-m") bUnique = false;
			else if (parameter == "-d" || parameter == "-debug") bDebugMode = true;
			else if (parameter == "-v" || parameter == "--version")
			{
				fprintf(stderr, "MapCaller v%s\n\n", VersionStr);
				exit(0);
			}
			else
			{
				fprintf(stderr, "Warning! Unknow parameter: %s\n", argv[i]);
				ShowProgramUsage(argv[0]);
				exit(0);
			}
		}
		if (bGVCF && bMonomorphic) bGVCF = false;
		if (iMaxDuplicate <= 0 || iMaxDuplicate > 15) iMaxDuplicate = 15;

		if (ReadFileNameVec1.size() == 0)
		{
			fprintf(stderr, "Warning! Please specify a valid read input!\n");
			ShowProgramUsage(argv[0]);
			exit(0);
		}
		if (ReadFileNameVec2.size() > 0 && ReadFileNameVec1.size() != ReadFileNameVec2.size())
		{
			fprintf(stderr, "Warning! Paired-end reads input numbers do not match!\n");
			fprintf(stderr, "Read1:\n"); for (vector<string>::iterator iter = ReadFileNameVec1.begin(); iter != ReadFileNameVec1.end(); iter++) fprintf(stderr, "\t%s\n", (char*)iter->c_str());
			fprintf(stderr, "Read2:\n"); for (vector<string>::iterator iter = ReadFileNameVec2.begin(); iter != ReadFileNameVec2.end(); iter++) fprintf(stderr, "\t%s\n", (char*)iter->c_str());
			exit(0);
		}
		if (CheckInputFiles(ReadFileNameVec1) == false || CheckInputFiles(ReadFileNameVec2) == false) exit(0);

		if (strcmp(LogFileName, "job.log")!= 0 && CheckOutputFileName(LogFileName) == false) exit(0);
		if (SamFileName != NULL && CheckOutputFileName(SamFileName) == false) exit(0);
		if (VcfFileName != NULL && CheckOutputFileName(VcfFileName) == false) exit(0);

		if (RefFileName != NULL)
		{
			random_prefix = MakeRefIdx(RefFileName);
			IndexFileName = (char*)random_prefix.c_str();
			bwa_idx_build(RefFileName, IndexFileName);
		}
		if (IndexFileName != NULL && CheckBWAIndexFiles(IndexFileName)) RefIdx = bwa_idx_load(IndexFileName);
		else
		{
			fprintf(stderr, "Warning! Please specify a valid reference index!\n");
			ShowProgramUsage(argv[0]);
			exit(0);
		}
		if (RefIdx == 0) fprintf(stderr, "\n\nError! Index files are corrupt!\n");
		else
		{
			Refbwt = RefIdx->bwt;
			RestoreReferenceInfo();
			//if (KnownSiteFileName != NULL) LoadKnownSites(KnownSiteFileName);

			if (GenomeSize <= 0)
			{
				fprintf(stderr, "Reference genome is empty\n");
				exit(1);
			}
			if (bVCFoutput)
			{
				fprintf(stderr, "Initialize the alignment profile...\n");
				MappingRecordArr = new MappingRecord_t[GenomeSize]();
			}
			pthread_mutex_init(&VarLock, NULL); pthread_mutex_init(&OutputLock, NULL); pthread_mutex_init(&LibraryLock, NULL); pthread_mutex_init(&ProfileLock, NULL);

			StartProcessTime = time(NULL);
			FILE *log = fopen(LogFileName, "a"); fprintf(log, "%s\n[CMD]", string().assign(80, '*').c_str()); for (i = 0; i < argc; i++) fprintf(log, " %s", argv[i]); fprintf(log, "\n\n"); fclose(log);

			Mapping();
			if (bVCFoutput) VariantCalling();

			bwa_idx_destroy(RefIdx);
			if (RefSequence != NULL) delete[] RefSequence;
			if (MappingRecordArr != NULL) delete[] MappingRecordArr;
			if (RefFileName != NULL)
			{
				random_prefix = "rm -f " + random_prefix + "*";
				system(random_prefix.c_str());
			}
			log = fopen(LogFileName, "a"); 
			fprintf(log, "All done! It took %lld seconds to complete the data analysis.\n\n\n", (long long)(time(NULL) - StartProcessTime)); fclose(log);
			fprintf(stderr, "All done! It took %lld seconds to complete the data analysis.\n", (long long)(time(NULL)- StartProcessTime));
		}
	}
	return 0;
}
